StandaloneMmPkg/Core/StandaloneMmCore.c

```diff
@@ -440,33 +500,77 @@ MmEntryPoint (
-      Status                      = MmiManage (
-                                      &CommunicateHeader->HeaderGuid,
-                                      NULL,
-                                      CommunicateHeader->Data,
-                                      (UINTN *)&gMmCorePrivate->BufferSize
-                                      );
+  if ((mMmCommunicationBuffer != NULL) && (mInternalCommBufferCopy != NULL)) {
+    CommunicationInOut = (COMMUNICATION_IN_OUT *)(UINTN)mMmCommunicationBuffer->CommunicationInOut;
+    if (CommunicationInOut->IsCommBufferValid) {
       //
-      // Update CommunicationBuffer, BufferSize and ReturnStatus
-      // Communicate service finished, reset the pointer to CommBuffer to NULL
+      // Synchronous MMI for MM Core or request from Communicate protocol
       //
-      gMmCorePrivate->BufferSize         += OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data);
-      gMmCorePrivate->CommunicationBuffer = 0;
-      gMmCorePrivate->ReturnStatus        = (Status == EFI_SUCCESS) ? EFI_SUCCESS : EFI_NOT_FOUND;
+      if (!MmIsBufferOutsideMmValid (
+             (UINTN)mMmCommunicationBuffer->FixedCommBuffer,
+             mMmCommunicationBuffer->FixedCommBufferSize
+             ))
+      {
+        //
+        // If CommunicationBuffer is not in valid address scope, return EFI_INVALID_PARAMETER
+        //
+        CommunicationInOut->ReturnStatus = EFI_INVALID_PARAMETER;
+      } else {
+        CommunicateHeader = (EFI_MM_COMMUNICATE_HEADER *)(UINTN)mMmCommunicationBuffer->FixedCommBuffer;
+        BufferSize        = OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data) + CommunicateHeader->MessageLength;
+        if (BufferSize <= mMmCommunicationBuffer->FixedCommBufferSize) {
+          //
+          // Shadow the communication buffer to internal buffer
+          //
+          ZeroMem (mInternalCommBufferCopy, mMmCommunicationBuffer->FixedCommBufferSize);
+          CopyMem (
+            mInternalCommBufferCopy,
+            (VOID *)(UINTN)mMmCommunicationBuffer->FixedCommBuffer,
+            BufferSize
+            );
+
+          CommunicateHeader = (EFI_MM_COMMUNICATE_HEADER *)mInternalCommBufferCopy;
+          BufferSize        = CommunicateHeader->MessageLength;
+          Status            = MmiManage (
+                                &CommunicateHeader->HeaderGuid,
+                                NULL,
+                                CommunicateHeader->Data,
+                                &BufferSize
+                                );
+
+          BufferSize = BufferSize + OFFSET_OF (EFI_MM_COMMUNICATE_HEADER, Data);
+          if (BufferSize <= mMmCommunicationBuffer->FixedCommBufferSize) {
+            //
+            // Copy the data back to FixedCommBuffer
+            //
+            ZeroMem (
+              (VOID *)(UINTN)mMmCommunicationBuffer->FixedCommBuffer,
+              mMmCommunicationBuffer->FixedCommBufferSize
+              );

```

Comment:
No need to call ZeroMem() again.
The purpose of ZeroMem() is to not pass data from one Handler to another.